packages/astro/src/content/runtime.ts

```diff
@@ -87,11 +88,39 @@ export function createGetCollection({
 				const data = rawEntry.filePath
 					? updateImageReferencesInData(rawEntry.data, rawEntry.filePath, imageAssetMap)
 					: rawEntry.data;
-				const entry = {
-					...rawEntry,
-					data,
-					collection,
-				};
+				let entry;
+				if (rawEntry.isModule === true) {
+					// @ts-expect-error	virtual module
+					const { default: contentModules } = await import('astro:content-module-imports');
+					const module = contentModules.get(rawEntry.filePath);
+					const resolvedComponent = await module();
+
+					entry = {
+						...rawEntry,
+						collection,
+						id: rawEntry.id,
+						slug: rawEntry.id,
+						body: rawEntry.body,
+						data: rawEntry.data,
+						async render() {
+							return {
+								Content: resolvedComponent.Content,
+								rendered: {
+									metadata: {
+										headings: resolvedComponent.getHeadings?.() ?? [],
+										remarkPluginFrontmatter: resolvedComponent.frontmatter ?? {},
+									}
+								}
+							};
+						},
+					};
```

Comment:
To match the markdown behaviour, this could be moved into the `renderEntry` function. That also means the module doesn't need to be resolved unless it's actually rendered.